kvserver: sync before removing sideloaded files

pav-kv opened the pull request:
A typical raft log truncation command removes log entries and updates the state machine in a single Pebble batch. This ensures that the state machine's `TruncatedState` is always up-to-date with the log.

However, if some of the truncated entries are stored in the sideloaded storage (typically `AddSST` commands), the corresponding files are removed separately from the Pebble batch (when applying side effects). If the batch is not synced before the files are removed, a process crash may cause an inconsistent state: the `TruncatedState` in the state machine
does not reflect the removal of the entries, but the entry files were removed. After restart, raft may try to load these entries in order to apply them or send to other replicas. It will find them missing and crash loop.

This commit syncs the state machine upon applying truncation commands that remove at least one sideloaded entry, to prevent the issue above.

Epic: none

Fixes #38566, #113135

Release note (bug fix): this commit fixes a bug in raft log truncation that could lead to crash loops, and unrecoverable loss of quorum in the unlikely worst case that all replicas enter this crash loop. The bug manifests when a few things coincide. The cluster is running a bulk write workload (e.g. schema change, import, restore), a log truncation command is running, and CRDB process crashes at an unfortunate moment (e.g. the process is killed, or kills itself for reasons like when it detects a disk stall).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114191)
<!-- Reviewable:end -->

pav-kv committed:
kvserver: document state machine sync nuances

Epic: none
Release note: none

pav-kv committed:
logstore: regression test for sideloaded storage panic

This commit adds a regression test for the issue fixed in the previous
commit. Without the commit, the test panics because the sideloaded
storage loses entries which the follower Replica will try to apply after
a restart.

For a detailed explanation of the scenario, see the previous commit and
the comment in the test.

Epic: none
Release note: none

pav-kv committed:
kvserver: sync before removing sideloaded files

A typical raft log truncation command removes log entries and updates
the state machine in a single Pebble batch. This ensures that the state
machine's TruncatedState is always up-to-date with the log.

However, if some of the truncated entries are stored in the sideloaded
storage (typically AddSST commands), the corresponding files are removed
separately from the Pebble batch (when applying side effects). If the
batch is not synced before the files are removed, a process crash may
cause an inconsistent state: the TruncatedState in the state machine
does not reflect the removal of the entries, but the entry files were
removed. After restart, raft may try to load these entries in order to
apply them or send to other replicas. It will find them missing and
crash loop.

This commit syncs the state machine upon applying truncation commands
that remove at least one sideloaded entry, to prevent the issue above.

Epic: none

Release note (bug fix): this commit fixes a bug in raft log truncation
that could lead to crash loops, and unrecoverable loss of quorum in the
unlikely worst case that all replicas enter this crash loop. The bug
manifests when a few things coincide. The cluster is running a bulk
write workload (e.g. schema change, import, restore), a log truncation
command is running, and CRDB process crashes at an unfortunate moment
(e.g. the process is killed, or kills itself for reasons like when it
detects a disk stall).

pavelkalinnikov commented:
Last commit now contains the fix.

pav-kv committed:
kvserver: intercept raft sync and post-apply

This commit adds a couple of interceptors into the raft proposal's
lifetime: one allows blocking the command application flow (without
blocking the entry commit flow), another allows intercepting commands
after applying them to storage and applying the in-memory side effects.

Both interceptors are useful in the next commits that reproduce a
problem in raft log's sideloaded storage.

Epic: none
Release note: none

pav-kv committed:
logstore: option to disable log append randomization

This commit adds a knob that allows tests to disable the randomization
which sometimes makes asynchronous log appends synchronous. This is
necessary for tests that rely on the async log appends.

Epic: none
Release note: none

pavelkalinnikov commented:
@erikgrinaker This is ready for a first pass review. Commit-by-commit is recommended.

pavelkalinnikov commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -410,12 +415,23 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
 				return errors.Wrap(err, "unable to handle truncated state")
 			}
 		} else {
+			// TODO(pavelkalinnikov): ensure that this truncation path makes sure the
```
@sumeerbhola Do you know if it's true already? Is the decoupled truncation also susceptible to the same problem that this PR fixes?

erikgrinaker commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -410,12 +415,23 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
 		// going to bother with a long-running migration.
 		apply := !looselyCoupledTruncation || res.RaftExpectedFirstIndex == 0
 		if apply {
+			// TODO(pavelkalinnikov): counting bytes may be unreliable because 0 bytes
+			// does not mean 0 files. Make the interface more robust.
```
I don't think this presents a correctness problem at present though, because SST files cannot have 0 size by definition, right?

erikgrinaker commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -410,12 +415,23 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
 		if apply {
+			// TODO(pavelkalinnikov): counting bytes may be unreliable because 0 bytes
+			// does not mean 0 files. Make the interface more robust.
+			if rmBytes, _, err := b.r.raftMu.sideloaded.BytesIfTruncatedFromTo(
```
Is the cost of these calls a concern? It seems to iterate over all sideloaded files regardless of the truncated range. In the common case I expect this to be cheap, as the filesystem metadata is cached, but what about very long Raft logs where we're only nibbling off a small head? Do we need a fast path to probe for a single sideloaded file in the given range?

pavelkalinnikov replied:
> Is the cost of these calls a concern? It seems to iterate over all sideloaded files regardless of the truncated range.

Slightly a concern, yes. Good news is that it scans only the files relevant to this range (they are in a separate directory). Bad news: for a log with many sideloaded entries we have to scan a long list here.

> Do we need a fast path to probe for a single sideloaded file in the given range?

The problem here is that we don't know which entries are sideloaded or not (the command doesn't say so at evaluation time, and we don't have any other metadata to tell). If the entries are sideloaded with high probability then probing the last few entries in the span would be a good fast path. If nothing found, fall back to a full scan.

Unfortunately, the most common case of log truncation doesn't have a single sideloaded file, so any kind of probing would be inefficient here, it must fall back to the scan.

Note here: we would be fine with false positives and sync, but we are not fine with false negatives. So we could have some sampling strategy and a `MANIFEST` file with a short list of spans that have >= 1 sideloaded entry. If the span intersects one of those, sync. That's not super backportable, is it?

pavelkalinnikov replied:
We could choose to always sync the log truncation. That would be a performance concern too, right?

pavelkalinnikov replied:
Unfortunately, we don't have an iterative `ls` via storage interface, there is only the `List` method that fetches all names.

pavelkalinnikov replied:
NB: the decoupled truncation path [here](https://github.com/cockroachdb/cockroach/blob/ab868241a425cdd86719173664ccd0b3594652a8/pkg/kv/kvserver/replica_app_batch.go#L419) similarly [calls](https://github.com/cockroachdb/cockroach/blob/ab868241a425cdd86719173664ccd0b3594652a8/pkg/kv/kvserver/raft_log_truncator.go#L329) this sideloaded storage `ls`, so we at least have a precedent. However, the decoupled truncations path has known some performance problems.

CC @sumeerbhola maybe you have some thoughts.

erikgrinaker replied:
Can we discover this from `EntryEncoding.IsSideloaded()`? I'm not intimately familiar with this, but I believe we can determine this simply by looking at the first byte of `ReplicatedCmd.Entry.Data` and its `Type`.

erikgrinaker reviewed (approved):
Nice work. LGTM modulo the performance question.

>The bug manifests when a few things coincide. The cluster is running a RESTORE workload, a log truncation command is running, and CRDB process crashes at an unfortunate moment (e.g. the process is killed, or kills itself for reasons like when it detects a disk stall).

The bug isn't limited to RESTORE, it affects any AddSSTable operation, i.e. any bulk write operation (imports, schema changes, etc).

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114191)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOZ_720wsOuOvGsy8m:-NlOZ_720wsOuOvGsy8n:bhvkllk) ([raw file](https://github.com/cockroachdb/cockroach/blob/10ae3b14ddc405d355e826e0568e9170e26fed14/pkg/kv/kvserver/replica_app_batch.go#L570)):*
> ```Go
> 	// separated, we must either sync here unconditionally upon log truncation, or
> 	// apply the side effects asynchronously when we are sure that the state
> 	// machine engine has synced the application of this command.
> ```

the "apply the side effects asynchronously" is exactly what setting `kv.raft_log.loosely_coupled_truncation.enabled` to true does. It does not do the truncation until a reader constructed using `storage.GuaranteedDurability` loads a `RangeAppliedState` that is large enough. Worth pointing to that in this comment.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 433 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1420528588:-NlP0Tsi5IJTAJkyUpLU:bj2hlc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed830f282ecce88496ee4f3e2b960219e80dd48a/pkg/kv/kvserver/replica_app_batch.go#L433)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

@sumeerbhola Do you know if it's true already? Is the decoupled truncation also susceptible to the same problem that this PR fixes?
</blockquote></details>

No, it isn't susceptible since it does not truncate until the raftAppliedIndex is durable up to truncation index.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 420 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422320021:-NlP1SeC-OncJMSgZ439:bd44w3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/replica_app_batch.go#L420)):*
> We could achieve a similar effect by dropping the scan early. `BytesIfTruncatedFromTo` scans the full thing, but we need to stop at the first file that matches.

I agree that if we go with this high-level approach we should not use `BytesIfTruncatedFromTo` and simply write a new function `DiskSideloadStorage.hasAtLeastOneSideLoadedFile(from, to)` -- that saves us a file stat too. We still have the cost of listing the files and parsing the indices out of the filenames.

If we want to be more radical and save this listing cost (now or in v24.1):
Loosely coupled truncation already solves this problem because it queues the truncation requests and applies them when the durable `RaftAppliedIndex` is advanced enough. We did not enable it because the flush based durability advancement is too delayed, and prevents us from truncating the raft log while the entries are still in the memtable (which is desirable for lower write amp). In hindsight, that aspect of the implementation was too forward looking (looking ahead to a separated raft log), which was a mistake.
We can use a different (low lag) signal for knowing the value of the durable RaftAppliedIndex, and then leverage most of the loosely coupled truncation work: `Replica.handleRaftReadyRaftMuLocked` can retrieve the value of `Replica.mu.state.RaftAppliedIndex` (rai), before it calls `LogStore.StoreEntries`. And if this `StoreEntries` call fsyncs the batch (which is the common case when appending log entries), we are guaranteed that up to rai is durable.

To use this signal, we can cut out the `durabilityAdvancedCallback()` stuff (which was shared across replicas) and instead do most of the per-replica logic in `raftLogTruncator.TryEnactTruncations` with the given durable rai. This truncation work could be done in `storeEntriesAndCommitBatch` synchronously (the truncation would be a batch that does not need to sync). If we don't want the latency hit of doing this synchronously in `storeEntriesAndCommitBatch` (though our current strongly coupled raft log truncation also happens in-band) we could do some async scheduling while ensuring that the previously scheduled truncation is done before doing the next one.
___
*[`pkg/kv/kvserver/testing_knobs.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOkxw95Ur9QznWYUdv:-NlOkxw95Ur9QznWYUdw:b-it1xcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a8a08662bccafb336d0b26cf36bf97bffa14a5f/pkg/kv/kvserver/testing_knobs.go#L76)):*
> ```Go
> 	//
> 	// If async log writes are enabled, this callback blocks state machine
> 	// application for the entire store, until it returns. Note that this does not
> ```

I understand this "blocks state machine application for the entire store" since we are blocking the `SyncWaiterLoop` and there is only one of them for a store.
But for the sync case (next para), why does it block the "entire store's raft processing" given that the synchronous wait happens for a replica? Is this some effect due to the shared raft scheduler and using up all the goroutines?
___
*[`pkg/kv/kvserver/logstore/logstore.go` line 250 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOe95tADk-Oxas1PF7:-NlOe95tADk-Oxas1PF8:b8e0tkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/logstore/logstore.go#L250)):*
> ```Go
> 		// using a blocking log sync. This is a rare case, so it's not worth
> 		// optimizing for.
> 		!overwriting &&
> ```

what about this overwriting case? Shouldn't we force `wantsSync` to true, since we should not delete the corresponding side-loaded entries before the overwrite is durable?
Is it not an issue because the overwritten entries will never be committed, so we will never need to read the side-loaded sstables if we deleted them prematurely? If yes, this is worth a code comment.
___
*[`pkg/kv/kvserver/client_raft_log_queue_test.go` line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOmUcrBpuabbIN9Dql:-NlOmUcrBpuabbIN9Dqm:bfb5gtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/31755a26d661a127749abe8e4d223088b9d37dc0/pkg/kv/kvserver/client_raft_log_queue_test.go#L234)):*
> ```Go
> 	t.Log("committed AddSSTs")
> 	_, lastIndex := info(leader, "leader")
> 	info(follower, "follower")
> ```

IIUC, the test is arranging things such that the AddSStable writes to the raft log (which will fsync the WAL) all predate the application, unlike the usual case where the application get interleaved with the raft log triggered fsyncs (making the applications also durable).
Won't the truncation (which is also proposed through raft) need to be fsynced before it is considered committed? And that would be make the application durable? Or has this test also managed to delay application of the original AddSSTables until after this truncation command is also durable in the raft log?


<!-- Sent from Reviewable.io -->

pav-kv committed:
logstore: add HasAnyEntry check to sideloaded storage

To decide whether log truncation requires sync, we need to check whether
there are any sideloaded entries in the removed log entries range. This
commit adds the corresponding helper method to the sideloaded storage.

The existing BytesIfTruncatedFromTo method could be used instead, but it
is more expensive: it fetches stats for all the files it scans, whereas
we just need an existence check, and we can terminate the scan early.

Epic: none
Release note: none

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114191)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 419 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422313363:-NlSpQOCBsMV2E5kEZ7u:bt1nlwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/replica_app_batch.go#L419)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I don't think this presents a correctness problem at present though, because SST files cannot have 0 size by definition, right?
</blockquote></details>

Yeah, seems safe. Sumeer made a good point though that `BytesIfTruncatedFromTo` not only scans the entire directory, but also unnecessarily calls `Stat` on every file. So I fixed this TODO by adding a new cheaper method to the sideloaded storage interface (see the new commit).
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 420 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422320021:-NlP8qX99eFZ_Hv8dGBX:b-cx8yun) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/replica_app_batch.go#L420)):*
Nice, thank you @sumeerbhola for the background.

> If we want to be more radical and save this listing cost (now or in v24.1):

I think we're looking for a small fix for now, because IIUC most (all?) of the clusters still use the legacy truncation mechanism. Removing the legacy mechanism though sounds good though, we should consider it for next releases. Do we have an issue for this? Let's put in it the suggested ideas how to make it without the sep-raft-log dependency.

@erikgrinaker
> Then again, I guess scanning the log is likely to be even more expensive than scanning the sideloaded storage in the common case

That's right, matches my understanding.

> I'm inclined to reduce the cost of this scan to the extent possible though.

Got rid of the unnecessary `Stat` calls that Sumeer mentioned.

> Might also be worth doing a quick microbenchmark to look at the cost with warm caches (the common case), across a few orders of magnitude.

Do you mean orders of magnitude for the number of files that `List` returns? And a microbenchmark would be for the `List` call? (or the entire `HasEntries` call most of which is the `List`)
___
*[`pkg/kv/kvserver/testing_knobs.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOkxw95Ur9QznWYUdv:-NlP7vKd-fb8m3Obj68d:b7gmdo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a8a08662bccafb336d0b26cf36bf97bffa14a5f/pkg/kv/kvserver/testing_knobs.go#L76)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I understand this "blocks state machine application for the entire store" since we are blocking the `SyncWaiterLoop` and there is only one of them for a store.
But for the sync case (next para), why does it block the "entire store's raft processing" given that the synchronous wait happens for a replica? Is this some effect due to the shared raft scheduler and using up all the goroutines?
</blockquote></details>

Ah, right, maybe the sync case blocks just one shard, I'll clarify that. This technicality is precisely why I noted "Not recommended to use" to this case. The test added in this commit only exercises the async log write path.
___
*[`pkg/kv/kvserver/client_raft_log_queue_test.go` line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOmUcrBpuabbIN9Dql:-NlSswso62R15cKT9ZZ4:b-cv7at) ([raw file](https://github.com/cockroachdb/cockroach/blob/31755a26d661a127749abe8e4d223088b9d37dc0/pkg/kv/kvserver/client_raft_log_queue_test.go#L234)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

IIUC, the test is arranging things such that the AddSStable writes to the raft log (which will fsync the WAL) all predate the application, unlike the usual case where the application get interleaved with the raft log triggered fsyncs (making the applications also durable).
Won't the truncation (which is also proposed through raft) need to be fsynced before it is considered committed? And that would be make the application durable? Or has this test also managed to delay application of the original AddSSTables until after this truncation command is also durable in the raft log?
</blockquote></details>

All correct. Thanks to the async log writes, we manage to pause the command application flow, commit and sync all the commands (AddSSTables and Truncate) before the application of a few last AddSSTables and the Truncate. When we set the application flow free, it commits one or a few batches without syncing. If nothing else syncs during this period, then the Truncate application is not durable.

This test panics in 100% cases locally, i.e. the repro works. It would not panic only if there is some sneaky sync (e.g. other range's Replica commits some entries), but I haven't had problems with this for the repro purposes.
___
*[`pkg/kv/kvserver/client_raft_log_queue_test.go` line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422295562:-NlSsXkd035prFBNQTrE:baz9l6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/client_raft_log_queue_test.go#L137)):*
> this will fail without the next commit, right? Consider either squashing this into the next commit, or move it after it, so that all commits pass.

Right, it panics without the fix. Moved the test commit after the fix.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -415,13 +415,9 @@ func (b *replicaAppBatch) runPostAddTriggersReplicaOnly(
-			); err != nil {
-				return errors.Wrap(err, "unable to count sideloaded files to remove")
-			} else if rmBytes != 0 {
+			if has, err := b.r.raftMu.sideloaded.HasEntries(ctx, 0, res.State.TruncatedState.Index); err != nil {
```
Might need to move this check to `handleTruncatedStateBelowRaftPreApply`, and check `HasEntries` in `[b.state.TruncatedState, res.State.TruncatedState.Index)` instead of `[0, res.State.TruncatedState.Index)`.

Counter-intuitively though, the check from 0 is more efficient because it may terminate more early. It may result in a false-positive sync though.

pavelkalinnikov replied:
Not moving into `handleTruncatedStateBelowRaftPreApply` because it is used by the loosely coupled truncations path, which already ensures syncs and doesn't need this extra check.

Moved `HasEntries` to be after `handleTruncatedStateBelowRaftPreApply` though, because the latter can withdraw the truncation command.

pav-kv reviewed (commented):
Addressed some comments. Still WIP, don't review again yet.

pavelkalinnikov commented on pkg/kv/kvserver/client_raft_log_queue_test.go:
```diff
@@ -123,3 +132,231 @@ func TestRaftLogQueue(t *testing.T) {
 }
+
+// TestRaftLogQueueSideloaded verifies that the raft log queue correctly
+// truncates the raft log when it has sideloaded entries.
```
All done.

pavelkalinnikov commented on pkg/kv/kvserver/client_raft_log_queue_test.go:
```diff
@@ -123,3 +132,231 @@ func TestRaftLogQueue(t *testing.T) {
+
+// TestRaftLogQueueSideloaded verifies that the raft log queue correctly
+// truncates the raft log when it has sideloaded entries.
+func TestRaftLogQueueSideloaded(t *testing.T) {
```
Renamed appropriately. Leaving this test in this file though, there is not "best" file for this.

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114191)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOZ_720wsOuOvGsy8m:-NlZW6RLD5bCbU6hy6JC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10ae3b14ddc405d355e826e0568e9170e26fed14/pkg/kv/kvserver/replica_app_batch.go#L570)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the "apply the side effects asynchronously" is exactly what setting `kv.raft_log.loosely_coupled_truncation.enabled` to true does. It does not do the truncation until a reader constructed using `storage.GuaranteedDurability` loads a `RangeAppliedState` that is large enough. Worth pointing to that in this comment.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/testing_knobs.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOkxw95Ur9QznWYUdv:-NlZWIuuAhvUBQLqqxkx:blnvrcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a8a08662bccafb336d0b26cf36bf97bffa14a5f/pkg/kv/kvserver/testing_knobs.go#L76)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Ah, right, maybe the sync case blocks just one shard, I'll clarify that. This technicality is precisely why I noted "Not recommended to use" to this case. The test added in this commit only exercises the async log write path.
</blockquote></details>

Clarified the blocking behaviour a bit, and added a TODO to make it simple.
___
*[`pkg/kv/kvserver/logstore/logstore.go` line 250 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOe95tADk-Oxas1PF7:-NlZWXj7E2BrNMKhyJOC:bb0rgaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/logstore/logstore.go#L250)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what about this overwriting case? Shouldn't we force `wantsSync` to true, since we should not delete the corresponding side-loaded entries before the overwrite is durable?
Is it not an issue because the overwritten entries will never be committed, so we will never need to read the side-loaded sstables if we deleted them prematurely? If yes, this is worth a code comment.
</blockquote></details>

Good point. This might be worth a separate issue and PR. I don't think it presents a correctness issue, but I think it might leave some files undeleted (we have an issue for that).

The `MaybeSideloadEntries` call above writes and syncs all the new sideloaded entries. The filename is based on (term, index), which uniquely identifies a raft entry. None of the entries will be removed, but for some indices we will have > 1 file (at different terms).

The following scenario then happens:
1. overwriting = true (i.e. there was a leader change, and the new leader shadows some of our entries in favour of higher term ones). Generally, entries from multiple previous terms could have been overridden.
2. Some of the overridden entries are sideloaded/AddSSTs. None of them were physically removed. We now have both old and new entries stored.
3. nonBlockingSync := false
4. batch.Commit(true) updates the HardState (now the metadata durably indicates that we're using the new files instead of the old files)
5. maybePurgeSideloaded removes the old files (safe now because the previous step synced)

If we crash in between steps, we're fine because during the transition period both old and new entry files are durable.

But the code below is incorrect in thinking that only term-1 files can be wiped. Also, it's possible that after a crash/restart (before step 4) some sideloaded files are now "too new". We won't read from them, but this is just another candidate for a cleanup after restart.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
@erikgrinaker @sumeerbhola All your comments are addressed, could you make another pass?

The only outstanding one (I think) is making sure that `FS.List()` calls on each truncation command don't kill performance. Even if we do a fast-path probe and `Stat` one entry before doing a full `List()`, that could be kind of expensive too.

A somewhat good fast-path that I can think of focuses on not doing any `FS` calls if we definitely know that the directory is empty:
- Maintain a bool saying whether the directory is empty.
- If we never called `FS.List()`, call it to initialize this bool (lazily, from the `HasAnyEntries()` method).
- Every sideloaded storage `Add()` will set this bool to true.
- Every sideloaded `TruncateTo()` call will reset it to false if all files were deleted (we already have a code that determines that).

How frequent are log truncation commands?

sumeerbhola reviewed (approved):
IIRC, our side-loaded files are typically > 1MB, which is by itself above `RaftLogQueueStaleSize` of 64KB, which allows us to truncate. So there should be only a few sideloaded files in the directory when we truncate. I imagine the List will be very fast in that case.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 3 files at r17, 3 of 3 files at r18, 3 of 5 files at r19, 2 of 2 files at r20, 1 of 1 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-:-Nl_-7gDAZ_RhxiVqJ21:b-3c4s7p)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/logstore/logstore.go` line 250 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-NlOe95tADk-Oxas1PF7:-NlZyjrE0iBUNHoMv0ql:b-4p7avb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/logstore/logstore.go#L250)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Good point. This might be worth a separate issue and PR. I don't think it presents a correctness issue, but I think it might leave some files undeleted (we have an issue for that).

The `MaybeSideloadEntries` call above writes and syncs all the new sideloaded entries. The filename is based on (term, index), which uniquely identifies a raft entry. None of the entries will be removed, but for some indices we will have > 1 file (at different terms).

The following scenario then happens:
1. overwriting = true (i.e. there was a leader change, and the new leader shadows some of our entries in favour of higher term ones). Generally, entries from multiple previous terms could have been overridden.
2. Some of the overridden entries are sideloaded/AddSSTs. None of them were physically removed. We now have both old and new entries stored.
3. nonBlockingSync := false
4. batch.Commit(true) updates the HardState (now the metadata durably indicates that we're using the new files instead of the old files)
5. maybePurgeSideloaded removes the old files (safe now because the previous step synced)

If we crash in between steps, we're fine because during the transition period both old and new entry files are durable.

But the code below is incorrect in thinking that only term-1 files can be wiped. Also, it's possible that after a crash/restart (before step 4) some sideloaded files are now "too new". We won't read from them, but this is just another candidate for a cleanup after restart.
</blockquote></details>

I was thinking of the above case and `willSync=false`, so we do `batch.Commit(false)`.

If there are potentially some problems here, could you create an issue so we don't forget about them?


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
The regression test hung in [one](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/13126093?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true&showLog=13126092_15468_275.282.24720.24721&logFilter=debug&logView=flowAware) of the CI runs. Apparently blocking the application flow for the entire store on one of the follower nodes can lead to a deadlock. Not sure what exactly deadlocks, but I suspect the reason is that there are some other ranges whose leader or leaseholder is collocated with the follower replica. For example, blocking application flow may block a lease transfer.

I will change the test in such a way that the range in question is the only range on the store.

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Let's clarify in the release note that this applies to all bulk writes (e.g. schema changes, imports, etc), not just restore.

Should we let this bake for a bit before backporting to supported releases? I think it's fine to leave this for 23.2.1, instead of rushing it into 23.2.0.

> Reviewed 1 of 2 files at r12, 1 of 3 files at r17, 3 of 3 files at r18, 3 of 5 files at r19, 2 of 2 files at r20, 1 of 1 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114191#-:-NlbpyeL2pbbwCJY5rTW:b-2k0rsp)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_app_batch.go` line 419 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422313363:-NlbpPvf2i8q8T28UQNx:b44tpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/replica_app_batch.go#L419)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Yeah, seems safe. Sumeer made a good point though that `BytesIfTruncatedFromTo` not only scans the entire directory, but also unnecessarily calls `Stat` on every file. So I fixed this TODO by adding a new cheaper method to the sideloaded storage interface (see the new commit).
</blockquote></details>

👍
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 420 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114191#gh-1422320021:-NlbpXFn4Jt1K5YkmzGO:bkam0nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0464d5fe1cfedc0784b7bd901dc5d07386ad7028/pkg/kv/kvserver/replica_app_batch.go#L420)):*
> Do you mean orders of magnitude for the number of files that `List` returns? And a microbenchmark would be for the `List` call? (or the entire `HasEntries` call most of which is the `List`)

I think eating the directory listing cost is fine for now, since we eliminated the rest of the cost. If it ever proves to be a problem, we can fix it then.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
I don't understand why the test still hangs on `WaitForVotersOrFatal`, and reliably does so only on Extended CI.

```
    testcluster.go:912:
          Error Trace:  github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:912
                              github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_log_queue_test.go:225
          Error:        Received unexpected error:
                        target store 3 not yet voter in range descriptor r68:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2LEARNER, (n3,s3):3LEARNER, next=4, gen=3, sticky=9223372036.854775807,2147483647]
                        (1) attached stack trace
                          -- stack trace:
                          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).waitForNewReplicas.func1
                          |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:956
                          | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
                          |   github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
                          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).waitForNewReplicas
                          |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:939
                          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).WaitForVoters
                          |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:904
                          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).WaitForVotersOrFatal
                          |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:912
                          | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCrashWhileTruncatingSideloadedEntries
                          |   github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_log_queue_test.go:225
                          | testing.tRunner
                          |   GOROOT/src/testing/testing.go:1595
                          | runtime.goexit
                          |   src/runtime/asm_amd64.s:1650
                        Wraps: (2) target store 3 not yet voter in range descriptor r68:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2LEARNER, (n3,s3):3LEARNER, next=4, gen=3, sticky=9223372036.854775807,2147483647]
                        Error types: (1) *withstack.withStack (2) *errutil.leafError
          Test:         TestCrashWhileTruncatingSideloadedEntries
```
I tried stressing it on Mac and GCE worker, it never gets stuck like this. Logs for successful and problematic runs in CI show no difference: the replication change after the `AddVotersOrFatal` call gets through.

pavelkalinnikov commented:
One flake in the added test is fixed, but [here](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_Tests_MaybeStressrace/13150454?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) is another (interesting) one.

```
F231214 23:42:22.059253 12595 go.etcd.io/raft/v3/log.go:322  [T1,Vsystem,n2,s2,r68/2:/{Table/Max-Max}] 1  tocommit(44) is out of range [lastIndex(43)]. Was the raft log corrupted, truncated, or lost?
```

Which means that the leader sent to a follower a commit index higher than the follower’s last index. But the leader never does this unless it [knows this index is durable](https://github.com/etcd-io/raft/blob/ee0fe9da492888b55fe183cf1a42931ad551ec6b/raft.go#L670-L676) on the follower.

In this test, the follower “crashed” and restarted, and lost a log entry at index 44. But why did the leader think this entry index is durable?

Since we emulate the crash, we `SetIgnoreSyncs(true)` and gracefully stop the cluster, then throw away all the side effects the node did after the `SetIgnoreSyncs(true)` call. But during this graceful shutdown, before it totally died, the node stored entry 44 (thinking that it synced it too) and sent a message to the leader accordingly.

To prevent this kind of anomaly we should /dev/null the communication of this node (i.e. throw away the communication side effects just as well as we throw away the storage side effects).

I will add some raft messages filtering to do that for now, but later will need to add a more general "stop all the communication" helper to the test node.

pavelkalinnikov commented:
The flakes are fixed now, good to merge.

pavelkalinnikov commented:
bors r=erikgrinaker,sumeerbhola
